lisa/sut_orchestrator/hyperv/platform_.py

```diff
@@ -334,3 +378,255 @@ def _delete_node(node_ctx: NodeContext, wait_delete: bool) -> None:
+                        device.instance_id = instance_id
+
+                        # add device to the given pool_type
+                        pool = self.AVAILABLE_HOST_DEVICES.get(pool_type, [])
+                        pool.append(device)
+                        self.AVAILABLE_HOST_DEVICES[pool_type] = pool
+
+    def _get_devices_from_pool(
+        self,
+        pool_type: HostDevicePoolType,
+        count: int,
+    ) -> List[DeviceAddressSchema]:
+        pool = self.AVAILABLE_HOST_DEVICES[pool_type]
+        if len(pool) < count:
+            raise SkippedException(
+                f"Not enough devices are available under pool: {pool_type}. "
+                f"Required count is {count}"
+            )
+        devices = pool[:count]
+
+        # Update the pool
+        pool = pool[count:]
+        self.AVAILABLE_HOST_DEVICES[pool_type] = pool
+
+        return devices
+
+    def _put_devices_into_pool(
+        self,
+        node_context: NodeContext,
+    ) -> None:
+        self._remove_devices_from_vm(node_context=node_context)
+
+        # Refresh the pool
+        self._configure_device_passthrough_pool(
+            self._hyperv_runbook.device_pools,
+            self.SUPPORTED_HOST_DEVICE_POOLTYPE,
+        )
+
+    def _remove_devices_from_vm(
+        self,
+        node_context: NodeContext,
+    ) -> None:
+        vm_name = node_context.vm_name
+        devices_ctx = node_context.device_passthrough_context
+        confing_commands = []
+        for ctx in devices_ctx:
+            for device in ctx.device_list:
+                confing_commands.append(
+                    f"Remove-VMAssignableDevice "
+                    f"-LocationPath '{device.location_path}' -VMName '{vm_name}'"
+                )
+                confing_commands.append(
+                    f"Mount-VMHostAssignableDevice -LocationPath "
+                    f"'{device.location_path}'"
+                )
+                confing_commands.append(
+                    f"Enable-PnpDevice -InstanceId '{device.instance_id}' "
+                    "-Confirm:$false"
+                )
+        self._run_powershell_commands(confing_commands)
+
+    def _assign_devices_to_vm(
+        self,
+        vm_name: str,
+        devices: List[DeviceAddressSchema],
+    ) -> None:
+        # Assign the devices to the VM
+        confing_commands = []
+        for device in devices:
+            confing_commands.append(
+                f"Disable-PnpDevice -InstanceId '{device.instance_id}' -Confirm:$false"
+            )
+            confing_commands.append(
+                f"Dismount-VMHostAssignableDevice -Force "
+                f"-LocationPath '{device.location_path}'"
+            )
+            confing_commands.append(
+                f"Add-VMAssignableDevice -LocationPath '{device.location_path}' "
+                f"-VMName '{vm_name}'"
+            )
+        self._run_powershell_commands(confing_commands)
+
+    def _run_powershell_commands(
```

Comment:
It looks not saving much lines, by running command in a batch. And it's hard to troubleshooting by lack of call trace and not check exit code.

Reply:
@squirrelsc 

Sure, I will remove `_run_powershell_commands()`

We call `run_cmdlet()` of PowerShell tool and by default it has `fail_on_error: bool = True`

Reply:
@squirrelsc Removed the fn now